agent/lib/kontena/network_adapters/weave.rb

```diff
@@ -1,104 +1,138 @@
   class Weave
     include Celluloid
     include Celluloid::Notifications
     include Kontena::Helpers::IfaceHelper
+    include Kontena::Helpers::LauncherHelper
     include Kontena::Helpers::WeaveHelper
     include Kontena::Logging
     include Kontena::Observer
+    include Kontena::Observable
 
+    WEAVEWAIT_NAME = "weavewait-#{WEAVE_VERSION}"
     DEFAULT_NETWORK = 'kontena'.freeze
 
-    finalizer :finalizer
-
-    def initialize(autostart = true)
-      @images_exist = false
-      @starting = false
-      @started = false
-
+    def initialize(start: true)
       info 'initialized'
-      subscribe('ipam:start', :on_ipam_start)
-      async.ensure_images if autostart
 
-      @ipam_client = IpamClient.new
-
-      # Default size of pool is number of CPU cores, 2 for 1 core machine
-      @executor_pool = WeaveExecutor.pool(args: [autostart])
+      async.start if start
+    end
 
-      async.start if autostart
+    # @return [Kontena::NetworkAdapters::IpamClient]
+    def ipam_client
+      @ipam_client ||= IpamClient.new
     end
 
     def start
-      observe(Actor[:node_info_worker]) do |node|
-        launch(node)
+      self.ensure_weavewait
+
+      observe(Actor[:node_info_worker], Actor[:weave_launcher], Actor[:ipam_plugin_launcher]) do |node, weave, ipam|
+        update(node) unless updated? # only once
       end
     end
 
-    def finalizer
-      @executor_pool.terminate if @executor_pool.alive?
-    rescue
-      # If Celluloid manages to terminate the pool (through GC or by explicit shutdown) it will raise
+    # @return [Boolean]
+    def updated?
+      !!@updated
     end
 
-    def api_client
-      @api_client ||= Excon.new("http://127.0.0.1:6784")
+    # @raise [Docker::Error]
+    # @return [Docker::Container]
+    def ensure_weavewait
+      unless container = inspect_container(WEAVEWAIT_NAME)
+        container = Docker::Container.create(
+          'name' => WEAVEWAIT_NAME,
+          'Image' => Kontena::NetworkAdapters::WeaveExecutor::IMAGE,
+          'Entrypoint' => ['/bin/false'],
+          'Labels' => {
+            'weavevolumes' => ''
+          },
+          'Volumes' => {
+            '/w' => {},
+            '/w-noop' => {},
+            '/w-nomcast' => {}
+          }
+        )
+      end
     end
 
-    def weave_api_ready?
-      # getting status should be pretty fast, set low timeouts to fail faster
-      response = api_client.get(path: '/status', :connect_timeout => 5, :read_timeout => 5)
-      response.status == 200
-    rescue Excon::Error
-      false
-    end
+    # @param node [Node]
+    def update(node)
+      state = self.ensure(node)
 
-    # @return [Boolean]
-    def running?
-      return false unless weave_container_running?
-      return false unless weave_api_ready?
-      return false unless interface_ip('weave')
-      true
+      update_observable(state)
+
+      @updated = true
+
+    rescue => exc
+      @updated = false
+
+      error exc
+
+      reset_observable
```

Comment:
Trapping errors here as `reset_observable` without letting the actor crash means that this actor will not actually retry on things like temporary IPAM errors from `ensure_default_pool`, if the kontena-etcd/kontena-ipam-plugin is slow to start up...